File: typescript/packages/core/src/core/mongodb/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { MongoDBAccessor } from '../../accessor/mongodb.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { PathSpec } from '../../types.ts'
import { databaseExists, entityExists } from './_client.ts'
import { buildCollectionSchemaJson, buildDatabaseJson } from './_schema_json.ts'
import { detectScope } from './scope.ts'
import { readStream, stringifyDoc } from './stream.ts'
import { ScopeLevel } from './types.ts'

function notFound(p: string): Error {
  const err = new Error(p) as Error & { code?: string }
  err.code = 'ENOENT'
  return err
}

export async function* streamAny(
  accessor: MongoDBAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): AsyncIterable<Uint8Array> {
  const scope = detectScope(path)
  if (scope.level === ScopeLevel.DOCUMENTS && scope.database !== null && scope.name !== null) {
    yield* readStream(accessor, path)
    return
  }
  yield await read(accessor, path, index)
}

export async function read(
  accessor: MongoDBAccessor,
  path: PathSpec | string,
  _index?: IndexCacheStore,
): Promise<Uint8Array> {
  const spec = typeof path === 'string' ? PathSpec.fromStrPath(path) : path
  const scope = detectScope(spec)

  if (scope.level === ScopeLevel.DOCUMENTS && scope.database !== null && scope.name !== null) {
    if (!(await entityExists(accessor, scope.database, scope.name, scope.kind))) {
      throw notFound(spec.original)
    }
    const chunks: Uint8Array[] = []
    let total = 0
    for await (const chunk of readStream(accessor, spec)) {
      chunks.push(chunk)
      total += chunk.byteLength
    }
    const buf = new Uint8Array(total)
    let off = 0
    for (const c of chunks) {
      buf.set(c, off)
      off += c.byteLength
    }
    return buf
  }

  if (scope.level === ScopeLevel.SCHEMA_JSON && scope.database !== null && scope.name !== null) {
    if (!(await entityExists(accessor, scope.database, scope.name, scope.kind))) {
      throw notFound(spec.original)
    }
    const payload = await buildCollectionSchemaJson(accessor, scope.database, scope.name)
    return new TextEncoder().encode(
      stringifyDoc(payload as unknown as Record<string, unknown>) + '\n',
    )
  }

  if (scope.level === ScopeLevel.DATABASE_JSON && scope.database !== null) {
    if (!(await databaseExists(accessor, scope.database))) throw notFound(spec.original)
    const payload = await buildDatabaseJson(accessor, scope.database)
    return new TextEncoder().encode(
      stringifyDoc(payload as unknown as Record<string, unknown>) + '\n',
    )
  }

  throw notFound(spec.original)
}

File: python/mirage/core/mongodb/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from bson.json_util import RELAXED_JSON_OPTIONS, dumps

from mirage.accessor.mongodb import MongoDBAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.mongodb._client import database_exists, entity_exists
from mirage.core.mongodb._schema_json import (build_collection_schema_json,
                                              build_database_json)
from mirage.core.mongodb.scope import detect_scope
from mirage.core.mongodb.stream import read_stream
from mirage.core.mongodb.types import ScopeLevel
from mirage.types import PathSpec
from mirage.utils.errors import enoent


async def read(
    accessor: MongoDBAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> bytes:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    scope = detect_scope(path)
    if scope.level == ScopeLevel.DOCUMENTS:
        if not await entity_exists(accessor.client, accessor.config,
                                   scope.database, scope.name, scope.kind,
                                   accessor):
            raise enoent(path)
        chunks: list[bytes] = []
        async for chunk in read_stream(accessor, path, index):
            chunks.append(chunk)
        return b"".join(chunks)
    if scope.level == ScopeLevel.SCHEMA_JSON:
        if not await entity_exists(accessor.client, accessor.config,
                                   scope.database, scope.name, scope.kind,
                                   accessor):
            raise enoent(path)
        payload = await build_collection_schema_json(accessor, scope.database,
                                                     scope.name)
        return (dumps(payload, json_options=RELAXED_JSON_OPTIONS) +
                "\n").encode()
    if scope.level == ScopeLevel.DATABASE_JSON:
        if not await database_exists(accessor.client, accessor.config,
                                     scope.database, accessor):
            raise enoent(path)
        payload = await build_database_json(accessor, scope.database)
        return (dumps(payload, json_options=RELAXED_JSON_OPTIONS) +
                "\n").encode()
    raise enoent(path)
